lib/context-manager.ts

```diff
@@ -0,0 +1,237 @@
+import { generateText, type ModelMessage } from 'ai';
+import { model } from '@/ai/providers';
+import { TOKEN_CONFIG } from './token-config';
+
+export function estimateTokens(content: string | ModelMessage[]): number {
+  if (typeof content === 'string') return Math.ceil(content.length / 3.5);
+
+  let totalChars = 0;
+  for (const msg of content) {
+    // Serialize the message
+    const msgStr = JSON.stringify(msg);
+    totalChars += msgStr.length;
+
+    //extra weight for tool invocations (they have significant overhead)
+    if ((msg as any).toolInvocations?.length) {
+      const toolCount = (msg as any).toolInvocations.length;
+      totalChars += toolCount * 50; // Approximate overhead per tool call
+    }
+  }
+
+  return Math.ceil(totalChars / 3.5);
+}
+
+export interface ContextManagerConfig {
+  maxTokens?: number;
+  keepRecentMessages?: number;
+  debug?: boolean;
+  systemPrompt?: string;
+}
+
+export const DEFAULT_CONTEXT_CONFIG: Required<
+  Omit<ContextManagerConfig, 'contextEditConfig' | 'systemPrompt'>
+> = {
```

Comment:
a bit confused here. `ContextManagerConfig` doesn't even have a `contextEditConfig` property that this is supposedly omitting. 

Also, I'm not sure why all properties of `ContextManagerConfig` are optional. Do they need to be? If not, let's forgo the optional `?` and that way this type simplifies to:
```suggestion
export const DEFAULT_CONTEXT_CONFIG: Omit<
  ContextManagerConfig, 'systemPrompt'>
> = {
```